rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+
+```json
+{
+  // New, complete updateEntityType message
+  "entityTypeId": "https://blockprotocol.org/@alice/entity-type/person",
+  "schema": {
+    "$id": "https://blockprotocol.org/@alice/entity-type/person/v/2",
+    "type": "object",
+    "kind": "entityType",
+    "title": "Person",
+    "properties": {
+      "https://blockprotocol.org/@alice/property-type/name/": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+      },
+      "https://blockprotocol.org/@alice/property-type/email": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/email/v/1"
+      },
+      "https://blockprotocol.org/@alice/property-type/phone-number": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
+      },
+      // The update is inlined in the existing schema
+      "https://blockprotocol.org/@alice/property-type/birth-date": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/birth-date/v/1"
+      }
+    }
+  }
+}
+```
+
+given that the original schema was created as follows
+
+```json
+{
+  "$id": "https://blockprotocol.org/@alice/entity-type/person/v/1",
+  "type": "object",
+  "kind": "entityType",
+  "title": "Person",
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/email": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/email/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/phone-number": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
+    }
+  }
+}
+```
+
+The above change applies to these existing BP operations:
+
+- `updateEntityType`
+- `updatePropertyType`
+- `updateLinkType`
+
+which must all make use of complete schemas in place of partial ones. These examples originate from the [Type System RFC](./0352-graph-type-system.md#interfacing-with-types-1).
+
+While not explained in detail in the Type System RFC the `aggregateEntities` operation would need to change behavior slightly. The types returned from aggregating on a specific entity type could now also be implicitly coerced instances of entities (supertypes projected from subtype instances).
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- The way this proposal adds type extension means that we must implement some version of property selection/projection for types, which comes with non-trivial implementation details for embedding applications.
+- This drifts further away from JSON Schema by introducing a different meaning to the `unevaluatedProperties` keyword and using it implicitly for our schemas.
+- The way compatibility is defined could result in newer versions of supertypes becoming incompatible over time. This is not unique to extending types and can be an "issue" in many parts of the type system as different versions of property types become incompatible.
+
```

Comment:
Additional drawback:
```suggestion
- When a new version of a supertype is published, new versions of the subtypes would have to be published to reference the new version of the supertype, cascading the update.

```

Reply:
I don't like that wording. New versions of the subtypes do not **have** to be published. I also don't think this is a drawback, it's just a characteristic of the approach. Another approach does not somehow fix this characteristic, so I don't think this is a drawback. 

Anyway if we do want to explicitly acknowledge it then lets be clear that this is about if a type _wants_ to stay updated with the things it references, then it will need to update each time they do.

Reply:
Agree that this drawback is more a consequence of references to types being versioned, rather than inheritance in itself, but still think it makes sense to mention somewhere - and unless we can think of a better place, I'd be in favour of mentioning it here.

Perhaps you'd prefer the following re-wording @Alfred-Mountfield ?
```suggestion

- When a new version of a supertype `A` is published, if a subtype `B` wants to extend the latest version of `A` a new version of subtype `B` would have to be published with an updated reference. This means updating supertypes can lead to cascading updates if the subtypes want to reference the latest version of the supertype.
```

Reply:
I do prefer that rewording, if you would like to mention it then I think that's fine. I still don't think this is a drawback, or even a negative thing. It just.. is _a_ thing?

As you say it's related to versioning, not the extension/duplication model, and even then I'm not sure cascading updates is really the right term, as these are (mostly) manual processes that won't happen rapidly, especially because we aren't using semver and won't be able to add something like `^1.0.0` or `~1.2.0`